valueformatting.go

```diff
@@ -0,0 +1,416 @@
+	},
+	"uint16": func(in []byte, byteOrder binary.ByteOrder) (string, error) {
+		v := make([]uint16, len(in)/2)
+		return binaryFormatterHelper(in, byteOrder, 2, &v)
+	},
+	"uint32": func(in []byte, byteOrder binary.ByteOrder) (string, error) {
+		v := make([]uint32, len(in)/4)
+		return binaryFormatterHelper(in, byteOrder, 4, &v)
+	},
+	"uint64": func(in []byte, byteOrder binary.ByteOrder) (string, error) {
+		v := make([]uint64, len(in)/8)
+		return binaryFormatterHelper(in, byteOrder, 8, &v)
+	},
+	"float32": func(in []byte, byteOrder binary.ByteOrder) (string, error) {
+		v := make([]float32, len(in)/4)
+		return binaryFormatterHelper(in, byteOrder, 4, &v)
+	},
+	"float64": func(in []byte, byteOrder binary.ByteOrder) (string, error) {
+		v := make([]float64, len(in)/8)
+		return binaryFormatterHelper(in, byteOrder, 8, &v)
+	},
+}
+
+func (formatting *valueFormatting) binaryFormatter(
+	encoding, ctype string,
+) valueFormatter {
+	var byteOrder binary.ByteOrder
+	// We don't check the get below because it's checked in
+	// validateType, which is called by validateFormat, which is
+	// called by format before calling this. :)
+	typeFormatter := binaryValueFormatters[ctype]
+	if encoding == "BigEndian" {
+		byteOrder = binary.BigEndian
+	} else {
+		byteOrder = binary.LittleEndian
+	}
+	return func(in []byte) (string, error) {
+		return typeFormatter(in, byteOrder)
+	}
+}
+
+func (formatting *valueFormatting) pbFormatter(ctype string) valueFormatter {
+	// We don't check the get below because it's checked in
+	// validateType, which is called by validateFormat, which is
+	// called by format before calling this. :)
+	md := formatting.pbMessageTypes[strings.ToLower(ctype)]
+	return func(in []byte) (string, error) {
+		message := dynamic.NewMessage(md)
+		err := message.Unmarshal(in)
+		if err == nil {
+			data, err := message.MarshalTextIndent()
+			if err == nil {
+				return string(data), nil
+			}
+		}
+		return "", err
+	}
+}
+
+var validValueFormattingEncodings = map[string]string{
+	"bigendian":       "BigEndian",
+	"b":               "BigEndian",
+	"binary":          "BigEndian",
+	"littleendian":    "LittleEndian",
+	"L":               "LittleEndian",
+	"hex":             "Hex",
+	"h":               "Hex",
+	"protocolbuffer":  "ProtocolBuffer",
+	"protocol-buffer": "ProtocolBuffer",
+	"protocol_buffer": "ProtocolBuffer",
+	"proto":           "ProtocolBuffer",
+	"p":               "ProtocolBuffer",
+	"":                "",
+}
+
+func (formatting *valueFormatting) validateEncoding(encoding string) (string, error) {
+	validEncoding, got := validValueFormattingEncodings[strings.ToLower(encoding)]
+	if !got {
+		return "", fmt.Errorf("invalid encoding: %s", encoding)
+	}
+	return validEncoding, nil
+}
+
+func (formatting *valueFormatting) validateType(
+	cname, validEncoding, encoding, ctype string,
+) (string, error) {
+	var got bool
+	switch validEncoding {
+	case "LittleEndian", "BigEndian":
```

Comment:
Having to switch on strings at this point seems a little janky, is there anyway we can have an enum of valid encoding types?

Reply:
Yes! Very janky. Fixed.